Source/WebDriver/WebSocketServer.cpp

```diff
@@ -0,0 +1,211 @@
+#include "Session.h"
+#include "WebDriverService.h"
+#include <algorithm>
+#include <optional>
+#include <wtf/JSONValues.h>
+#include <wtf/UUID.h>
+#include <wtf/text/MakeString.h>
+#include <wtf/text/WTFString.h>
+
+namespace WebDriver {
+
+WebSocketServer::WebSocketServer(WebSocketMessageHandler& messageHandler, WebDriverService& service)
+    : m_messageHandler(messageHandler)
+    , m_service(service)
+{
+}
+
+void WebSocketServer::addStaticConnection(WebSocketMessageHandler::Connection connection)
+{
+    m_staticConnections.push_back(connection);
+}
+
+void WebSocketServer::addConnection(std::pair<WebSocketMessageHandler::Connection, String> connectionAndSession)
+{
+    m_connectionToSession.push_back(connectionAndSession);
+}
+
+bool WebSocketServer::isStaticConnection(WebSocketMessageHandler::Connection& connection)
+{
+    return std::count(m_staticConnections.begin(), m_staticConnections.end(), connection);
+}
+
+void WebSocketServer::removeStaticConnection(WebSocketMessageHandler::Connection& connection)
+{
+    m_staticConnections.erase(std::find(m_staticConnections.begin(), m_staticConnections.end(), connection));
+}
+
+void WebSocketServer::removeConnection(WebSocketMessageHandler::Connection& connection)
+{
+    std::erase_if(m_connectionToSession, [&](const auto& pair) {
+        return pair.first == connection;
+    });
+}
+
+std::optional<RefPtr<Session>> WebSocketServer::getSession(WebSocketMessageHandler::Connection& connection)
+{
+    String sessionId = nullString();
+
+    for (const auto& pair : m_connectionToSession) {
+        if (pair.first == connection) {
+            sessionId = pair.second;
+            break;
+        }
+    }
+
+    if (sessionId.isNull())
+        return std::nullopt;
+
+    auto existingSession = m_service.session();
+    if (!existingSession || !(*existingSession) || (*existingSession)->id() != sessionId)
+        return std::nullopt;
+
+    return existingSession;
+}
+
+std::optional<WebSocketMessageHandler::Connection> WebSocketServer::getConnection(const String& sessionId)
+{
+    for (const auto& pair : m_connectionToSession) {
+        if (pair.second == sessionId)
+            return pair.first;
+    }
+
+    return std::nullopt;
+}
+
+String WebSocketServer::getSessionID(const String& resource)
+{
+    // https://w3c.github.io/webdriver-bidi/#get-a-session-id-for-a-websocket-resource
+
+    constexpr auto sessionPrefix = "/session/"_s;
+    if (!resource.startsWith(sessionPrefix))
+        return nullString();
+
+    auto sessionId = resource.substring(sessionPrefix.length());
+    auto uuid = WTF::UUID::parse(sessionId);
+    if (!uuid || !uuid->isValid())
+        return nullString();
+
+    return sessionId;
+}
+
+RefPtr<WebSocketListener> WebSocketServer::startListening(String sessionId)
+{
+    // https://w3c.github.io/webdriver-bidi/#start-listening-for-a-websocket-connection
+    // FIXME Add more listeners when start supporting multiple sessions
+    auto listener = m_listener;
```

Comment:
Why do we need the local variable?

Reply:
Removed. Might have been a leftover from following the spec too closely.